src/SignalR/clients/ts/signalr/src/HttpConnection.ts

```diff
@@ -501,3 +502,70 @@ export class HttpConnection implements IConnection {
 function transportMatches(requestedTransport: HttpTransportType | undefined, actualTransport: HttpTransportType) {
     return !requestedTransport || ((actualTransport & requestedTransport) !== 0);
 }
+
+export class TransportSendQueue {
+    private sending: boolean = false;
+    private buffer: any[] = [];
+    private promiseQueue?: Promise<void>;
+
+    public async send(transport: ITransport, data: string | ArrayBuffer): Promise<void> {
+        return this.sending ?
+            this.bufferData(transport, data) :
+            this.sendCore(transport, data);
+    }
+
+    private async sendCore(transport: ITransport, data: string | ArrayBuffer) {
+        this.sending = true;
+
+        let resolve: () => void;
+        this.promiseQueue = new Promise((r) => resolve = r);
+
+        try {
+            await transport.send(data);
+        } finally {
+            this.sending = false;
+            this.promiseQueue = undefined;
+            resolve!();
```

Comment:
The next call to sendBufferedData won't be called inline. Effectively it's called with a 0 setTimeout. If there's an intervening call to TransportSendQueue.send() it will see that this.sending is false and allow concurrent/out-of-order data.

The fix for this is usually to unset the flag (if necessary since more data could have been buffered while you were being scheduled) after the continuation awaiting the promise starts. In this case that would be sendBufferedData, but it's possible that there's no queued call to sendBufferedData.

That's one reason writing the async loop might be cleaner.